tests/test_yask.py

```diff
@@ -293,51 +307,102 @@ class TestOperatorRealAcoustic(object):
         rec.coordinates.data[:, 0] = np.linspace(0., model.domain_size[0], num=nrec)
         rec.coordinates.data[:, 1:] = src.coordinates.data[0, 1:]
+        return rec
 
-        # Create solver object to provide relevant operators
-        solver = AcousticWaveSolver(model, source=src, receiver=rec,
-                                    time_order=time_order,
-                                    space_order=space_order)
-
-        # Create adjoint receiver symbol
-        srca = Receiver(name='srca', ntime=solver.source.nt,
-                        coordinates=solver.source.coordinates.data)
-
-        # Run forward and adjoint operators
-        rec, _, _ = solver.forward(save=False)
-        solver.adjoint(rec=rec, srca=srca)
-
-        # Adjoint test: Verify <Ax,y> matches  <x, A^Ty> closely
-        term1 = np.dot(srca.data.reshape(-1), solver.source.data)
-        term2 = np.linalg.norm(rec.data) ** 2
-        info('<Ax,y>: %f, <x, A^Ty>: %f, difference: %12.12f, ratio: %f'
-             % (term1, term2, term1 - term2, term1 / term2))
-        assert np.isclose(term1, term2, rtol=1.e-5)
+    @pytest.fixture
+    def subs(self, model, u):
+        dt = model.critical_dt
+        return dict([(t.spacing, dt)] + [(time.spacing, dt)] +
+                    [(i.spacing, model.get_spacing()[j]) for i, j
+                     in zip(u.indices[1:], range(len(model.shape)))])
+
+    def test_acoustic_wo_src_wo_rec(self, model, stencil, subs, m, damp, u):
+        """
+        Test that the acoustic wave equation runs without crashing in absence
+        of sources and receivers.
+        """
+        u.data[:] = 0.0
+        eqn = [Eq(u.forward, stencil)]
+        op = Operator(eqn, subs=subs)
+        op.apply(u=u, m=m, damp=damp, t=10)
+
+    def test_acoustic_w_src_wo_rec(self, model, stencil, subs, m, damp, u, src):
+        """
+        Test that the acoustic wave equation runs without crashing in absence
+        of receivers.
+        """
+        dt = model.critical_dt
+        u.data[:] = 0.0
+        eqns = [Eq(u.forward, stencil)]
+        eqns += src.inject(field=u.forward, expr=src * dt**2 / m, offset=model.nbpml)
+        op = Operator(eqns, subs=subs)
+        op.apply(u=u, m=m, damp=damp, src=src, t=1)
+
+    def test_acoustic_w_src_w_rec(self, model, stencil, subs, m, damp, u, src, rec):
+        """
+        Test that the acoustic wave equation forward operator produces the correct
+        results when running the same model as in ``test_adjointA.py``.
+        """
+        dt = model.critical_dt
+        u.data[:] = 0.0
+        eqns = [Eq(u.forward, stencil)]
+        eqns += src.inject(field=u.forward, expr=src * dt**2 / m, offset=model.nbpml)
+        eqns += rec.interpolate(expr=u, offset=model.nbpml)
+        op = Operator(eqns, subs=subs)
+        op.apply(u=u, m=m, damp=damp, src=src, rec=rec, t=1)
+        # TODO: come up with proper asserts, but u.data looks sane already
```

Comment:
The assert could just be a known value of the L2 norm of `u` associated with the model/damp/.... setup

Reply:
> # TODO: come up with proper asserts, but u.data looks sane already
with t=1? everything will still be zero in `u` so not much to assert on. 

One possibility would be to assert against non-yask solution.
As it looks like everything is nicely setup, you could also run adjointTest on a single space_orde/time_order combination.

Reply:
I cannot run adjoint yet because of fix required in YASK

> That would work too for more time step but would restrict it to a the part of the domain where we are sure both YASK and non-YASK have solution value (remove PML part for example)

I can think of something along these lines

Reply:
BTW, to check correctness I was doing exactly that (only using 1-norm instead of 2-norm). Not sure why I initially opted to leave this kind of check out. Now implemented using 2-norms and comparing to the values obtained from an analogous test from test_adjointA